services/ymax-planner/test/blockchain-parsing.test.ts

```diff
@@ -0,0 +1,646 @@
+    },
+  ];
+
+  await processPendingTxEvents(events, mockHandler, createMockPendingTxOpts());
+
+  // Should only process the last value (index -1)
+  t.is(handledTxs.length, 1);
+  t.is(handledTxs[0]?.type, TxType.GMP);
+  t.is(handledTxs[0]?.amount, 2000000n);
+});
+
+// ============================================================================
+// 3. makeVstorageEvent Function Tests
+// ============================================================================
+
+test('makeVstorageEvent creates properly formatted event', t => {
+  const blockHeight = 123n;
+  const path = 'published.test.portfolios.portfolio1';
+  const value = { test: 'data', amount: AmountMath.make(depositBrand, 1000n) };
+
+  const { event, streamCellJson } = makeVstorageEvent(
+    blockHeight,
+    path,
+    value,
+    marshaller,
+  );
+
+  // Check event structure
+  t.is(event.type, 'state_change');
+  t.true(Array.isArray(event.attributes));
+  t.true(event.attributes!.length > 0);
+
+  // Check event attributes
+  const attrs = Object.fromEntries(
+    event.attributes!.map(a => [a.key, a.value]),
+  );
+  t.is(attrs.store, 'vstorage');
+  t.truthy(attrs.key);
+  t.truthy(attrs.value);
+
+  // Check stream cell JSON
+  const streamCell = JSON.parse(streamCellJson);
+  t.is(streamCell.blockHeight, '123');
+  t.true(Array.isArray(streamCell.values));
+  t.is(streamCell.values.length, 1);
+
+  // Check that value can be unmarshalled
+  const capData = JSON.parse(streamCell.values[0]);
+  const unmarshalled = marshaller.fromCapData(capData);
+  t.deepEqual(unmarshalled, value);
+});
+
+test('makeVstorageEvent handles nested path segments', t => {
+  const path = 'published.ymax1.portfolios.portfolio5.flows.flow3';
+  const value = { status: 'in-progress' };
+
+  const { event } = makeVstorageEvent(0n, path, value, marshaller);
+
+  const attrs = Object.fromEntries(
+    event.attributes!.map(a => [a.key, a.value]),
+  );
+
+  // Key should encode all path segments
+  const keyParts = attrs.key.split('\x00');
+  t.true(keyParts.length > 1);
+  // First part is the number of segments
+  t.is(keyParts[0], String(path.split('.').length));
+});
+
+test('makeVstorageEvent preserves complex nested structures', t => {
+  const complexValue = {
+    nested: {
+      deeply: {
+        value: 123,
+        array: [1, 2, 3],
+      },
+    },
+    amounts: [
+      AmountMath.make(depositBrand, 100n),
+      AmountMath.make(depositBrand, 200n),
+    ],
+    map: {
+      key1: 'value1',
+      key2: 'value2',
+    },
+  };
+
+  const { streamCellJson } = makeVstorageEvent(
+    0n,
+    'test.path',
+    complexValue,
+    marshaller,
+  );
+
+  const streamCell = JSON.parse(streamCellJson);
+  const capData = JSON.parse(streamCell.values[0]);
+  const unmarshalled = marshaller.fromCapData(capData);
+
+  t.deepEqual(unmarshalled, complexValue);
+});
+
+// ============================================================================
+// 4. EVM Log Parsing Edge Cases
```

Comment:
I don't think these are useful, but that's more a question for @frazarshad and @rabi-siddique. But if they do remain, I think they belong in their own file.